### PR TITLE
feat(#3097): propagate translation failures for gateway listeners referred secrets

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

It adds propagation of translation failures for Gateway.Listeners that:
* specify more than one certReference,
* refer to a secret that has no valid TLS key-pair embedded 

**Which issue this PR fixes**:

Part of #3097. 

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/parser/parser.go [hunk 1]
```diff
@@ -429,22 +431,17 @@ func getGatewayCerts(log logrus.FieldLogger, s store.Storer) []certWrapper {
...
 
 					// SecretObjectReference is a misnomer; it can reference any Group+Kind, but defaults to Secret
 					if ref.Kind != nil {
-						if string(*ref.Kind) != "Secret" {
-							log.WithFields(logrus.Fields{
-								"gateway":  gateway.Name,
-								"listener": listener.Name,
-							}).Error("CertificateRefs to kinds other than Secret are not supported")
+						if kind := *ref.Kind; kind != "Secret" {
+							p.registerTranslationFailure(fmt.Sprintf("CertificateRefs to kinds other than Secret (%s) are not supported", kind), gateway)
```
Comment (czeslavo) [inline internal/dataplane/parser/parser.go:443]: This is handled on the controller side (https://github.com/Kong/kubernetes-ingress-controller/blob/ec0f2702b99f8a75581e5c789bdf57a062ad1491/internal/controllers/gateway/gateway_utils.go#L462). Not possible to trigger this in the Parser.

File: internal/dataplane/parser/parser.go [hunk 2]
```diff
@@ -461,13 +458,8 @@ func getGatewayCerts(log logrus.FieldLogger, s store.Storer) []certWrapper {
...
 						if !newRefChecker(ref).IsRefAllowedByGrant(allowed) {
-							log.WithFields(logrus.Fields{
-								"gateway":           gateway.Name,
-								"gateway_namespace": gateway.Namespace,
-								"listener":          listener.Name,
-								"secret_name":       string(ref.Name),
-								"secret_namespace":  namespace,
-							}).WithError(err).Error("secret reference not allowed by ReferenceGrant")
+							secretName := namespace + "/" + string(ref.Name)
+							p.registerTranslationFailure(fmt.Sprintf("secret reference (%s) not allowed by ReferenceGrant", secretName), gateway)
```
Comment (czeslavo) [inline internal/dataplane/parser/parser.go:462]: Ditto, handled on the controller side:
https://github.com/Kong/kubernetes-ingress-controller/blob/ec0f2702b99f8a75581e5c789bdf57a062ad1491/internal/controllers/gateway/gateway_utils.go#L442

File: internal/dataplane/parser/parser.go (hunk 1 above)
Comment (rainest) [reply internal/dataplane/parser/parser.go:445]: To confirm then, this and the other one that's handled controller-side should just be removed? Not sure why we'd keep the parser-side code and convert it to the registration system if it's effectively unreachable.

Comment (czeslavo) [reply internal/dataplane/parser/parser.go]: Yeah, it seems to me we can drop both validations on the parser side. The logic verifying the reference permissions in the parser (`!newRefChecker(ref).IsRefAllowedByGrant(allowed)`) is duplicated in the `getReferenceGrantConditionReason` on the controller side: [e9bbb3a](https://github.com/Kong/kubernetes-ingress-controller/pull/3147/commits/e9bbb3a2a3850bda7ee17ee0d69abd58193bdd0c)

I've "moved" two unit tests that I found to be missing on the controller side from the parser suite (
8dc2ac6). Besides that, the ReferenceGrants compatibility is verified in the gateway conformance tests.

**Review by rainest — APPROVED** [review]

**Pull request merged**